src/software/ai/hl/stp/tactic/shadow_enemy/shadow_enemy_fsm.h

```diff
@@ -0,0 +1,190 @@
+        if(friendlyTeam.goalie().has_value()){
+            robots_to_ignore.emplace_back(friendlyTeam.goalie().value());
+        }
+
+        auto best_enemy_shot_opt = calcBestShotOnGoal(field,friendlyTeam,enemyTeam,shadowee.position(),TeamType::FRIENDLY, robots_to_ignore);
+
+        Vector enemy_shot_vector = Vector(0, 0);
+        if (best_enemy_shot_opt){
+            enemy_shot_vector = best_enemy_shot_opt->getPointToShootAt() - shadowee.position();
+        }else{
+            enemy_shot_vector = field.friendlyGoalCenter() - shadowee.position();
+        }
+        return shadowee.position() + enemy_shot_vector.normalize(shadow_distance);
+    }
+
+    auto operator()()
+    {
+        using namespace boost::sml;
+
+        const auto block_s          = state<BlockState>;
+        const auto steal_and_chip_s = state<StealAndChipState>;
+
+        const auto update_e = event<Update>;
+
+        /**
+         * Guard that checks if the enemy threat has ball
+         *
+         * @param event ShadowEnemyFSM::Update
+         *
+         * @return if the ball has been have_possession
+         */
+        const auto enemy_threat_has_ball = [](auto event) {
+            std::optional<EnemyThreat> enemy_threat_opt =
+                event.control_params.enemy_threat;
+            if (enemy_threat_opt.has_value())
+            {
+                return enemy_threat_opt.value().has_ball;
+            };
+            return false;
+        };
+
+        /**
+         * Guard that checks if the enemy threat is near
+         * our robot
+         *
+         * @param event ShadowEnemyFSM::Update
+         *
+         * @return if the ball has been have_possession
+         */
+        const auto enemy_threat_is_near = [](auto event) {
+            std::optional<EnemyThreat> enemy_threat_opt =
+                event.control_params.enemy_threat;
+            if (enemy_threat_opt.has_value())
+            {
+                return distance(enemy_threat_opt.value().robot.position(),
+                                event.common.robot.position()) <=
+                       event.control_params.steal_and_chip_distance;
+            };
+            return false;
+        };
+
+        /**
+         * Action to block the ball
+         *
+         * If the enemy can pass to the robot we are shadowing, block the pass.
+         * Else, block the net
+         *
+         * @param event ShadowEnemyFSM::Update
+         */
+        const auto block_pass_or_net = [](auto event) {
+            auto ball_position = event.common.world.ball().position();
+            auto face_ball_orientation =
+                (ball_position - event.common.robot.position()).orientation();
+            Point position_to_block;
+            if (!event.control_params.enemy_threat.value().has_ball)
+            {
+                position_to_block = findBlockPassPoint(ball_position, event.control_params.enemy_threat.value().robot,
```

Comment:
just a though: if you were shadowing close to the friendly net, would it be more safe to try to block the shot rather than try to intercept a pass?

Reply:
I think ideally we would never want to block the shot of the enemy robot that is passing to the robot we are shadowing as it should be the other shadower's job to block the net.